### PR TITLE
chore(tests): silence console.{log|warn|error} messages during tests

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
this adds a little utility for replacing the `console` functions with
mocks to silence any error messages and get our jest output under
control.

This function calls `afterAll` internally, so it will clean up after
itself by restoring the original functions for `console.log`, `.error`,
etc.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [ ] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe): testing quality-of-life improvement

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

Gaze upon this and feel at peace:

https://user-images.githubusercontent.com/6207644/174887868-ec9d7eb7-6862-471a-8cf0-91fec3ebdfa1.mov

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

This is just a testing change anyhow, so I think if tests are passing we are all good.

I did confirm that this setup does not inadvertently silence tests outside of the suites in which it's called.

### CODE REVIEW

**Review by rwaskiewicz — COMMENTED** [review]
The code/PR itself looks good to me.

I am a little concerned that we're suppressing error messages that we should be addressing in the test suites themselves though - 

For example, the log output from `src/compiler/style/test/optimize-css.spec.ts` is a consequence of how Autoprefixer is cached/installed. We have a task in the backlog to fix this (STENCIL-314). In fact, I managed to fix this once (STENCIL-131), but couldn't tell you for the life of me how it happened 😨 Unfortunately, not seeing that output is the best Acceptance Criteria that we have ATM to verify that we've fixed the caching of certain dependencies. As a result, I'm a little hesitant to suppress the output for this file.

Similarly, for the tests found in the `src/runtime/test` directory, I wonder how many of those tests can be fixed/reworked such that the original intent of the tests can be captured, but not require us to test in such a way that causes these errors to log.  Do we know if that's possible for those suites?

**Comment by alicewriteswrongs** [comment]
thanks for that context, I ran across the code in `src/sys/node/bundles/autoprefixer.js` just today and was scratching my head as to what exactly is going on there.

I think at the minimum it might be nice to silence some of the 'prop changed' warnings which some tests spew and also use the log mocking stuff in files like `terminal-logger.spec.ts` where we actually need to make assertions about logs, and then leave addressing the autoprefixer stuff for another day. I'll give it a go for the `src/runtime/test` ones and see what I can come up with without spending too much time on it

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz ok just updated this to take a different strategy for the tests which are doing `console.warn`, basically I went through and for some I was able to figure out a small change to the test component to resolve the warning messages, and for those which I wasn't able to fix I added a little wrapper function called `withSilentWarn` which takes a callback, silences `console.warn`, `await`s the callback, unsilences `console.warn`, and returns whatever was returned by the callback. This is still silencing these messages, but in a more specific and targeted way.

I also removed the changes for the autoprefixer bit

File: src/testing/testing-utils.ts [hunk 1]
```diff
@@ -95,3 +95,99 @@ function getAppUrl(config: d.Config, browserUrl: string, appFileName: string) {
...
+
+interface ConsoleMocker {
+  (): {
+    logMock: jest.Mock<typeof console.log>;
+    warnMock: jest.Mock<typeof console.warn>;
+    errorMock: jest.Mock<typeof console.error>;
+  };
+}
+
+// export function withSilentWarn<T extends () => Promise<any>>(
```
Comment (rwaskiewicz) [inline src/testing/testing-utils.ts:168]: Do we need this still? If not, can we remove it?

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
LGTM, two tiny nits

File: src/testing/testing-utils.ts (hunk 1 above)
Comment (alicewriteswrongs) [reply src/testing/testing-utils.ts:168]: oops 🤦‍♀️

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**